README.md
# Cocodrilo

This is the base application used at [CROCODΞ](http://www.crocode.mobi/). It uses **Rails 5** to manage the backend.
The purpose of this project is to:

- Provide a minimal and well-configured application generator.
- Improve some Rails defaults.
- Enforce **code style guidelines**.
- Provide _basic_ **security**.

This template does **NOT** aim to:

- Encourage having too many dependencies. Dependencies are not cheap, and each
  bundled tool must have a _good_ reason to be included.
- Provide "pattern" gems (e.g. decorator). Patterns are usually
  project-specific and most of them can be achieved **without** any gems or
  libraries.
- Be a silver bullet. This is tailored for the majority of Rails applications
  you may want to build, but we know it won't work sometimes.

## Pre-requisites

- [Ruby](https://www.ruby-lang.org) >= 2.3.1
- [Bundler](http://bundler.io/)
- [rvm](https://github.com/rvm/rvm)
- [Git](https://git-scm.com/)
- [PostgreSQL](https://www.postgresql.org/)
- [Redis](http://redis.io/) - primarily for use with [Sidekiq](http://sidekiq.org/)

## Installation

Install the gem:

```sh
gem install cocodrilo
```

This will make the `cocodrilo` command accessible throughout your `PATH`.

## Configuring your environment

### PostgreSQL

**development** and **test** databases are automatically setup for you.
While running the generator it's assumed that:

- The PostgreSQL server is running.
- You have a PostgreSQL user named after your UNIX login.
- Your PostgreSQL user has a _blank_ password.

It's OK if your PG settings happen to be different from this; DB creation will
fail, but you can do it manually thereafter.

At a basic level, here's how to setup PostgreSQL on Linux:

```sh
# Creates a user
sudo -u postgres createuser -s my_user_name

# Runs psql
sudo -u postgres psql

# Change your use password within psql. Leave it blank.
[local] local@user=# \password my_user_name
```

## Generating your app

You can generate a new app with the following command:

```sh
cocodrilo myapp
```

## Starting up your app

```sh
$ cd my_app_folder
$ foreman start
```

This command runs Rails server and Sidekiq all at once.
Note that Your redis server has to be up and running because of Sidekiq. To
customize these processes you can edit `Procfile`.

Now you can work as you'd usually work in any Rails application, with automatic
Ruby and JS file reloads out of the box.

## Basic hands-on guide

### App setup

Your team can use the following command after cloning the git repository:

```sh
bin/setup
```
### Deploy

Deploy your app with the following command:

```sh
# Replace `MY_ENV` with `production` or `staging`.
bin/deploy MY_ENV
```

This command pushes your code to Heroku, migrates your database and restarts
your dynos. It will work out of the box if you've generated your app with
`--heroku true`. If not, please create `staging` and `production` git remotes
pointing to the respective Heroku remotes.

### Health check

Check your app's health with the `rake health` command. It runs the following
tasks:

- `rspec` runs your Ruby and Rails specs.
- If you app happens to be below 90% test coverage the `rspec` command will fail.
- `bundle-audit` and `brakeman` check if your app does not have basic
  security holes.
- `rubocop` makes sure your code adheres to style guidelines.

### Testing

#### Ruby tests

Use the following command to run all your specs:

```sh
rspec
```

We use the following tools:

- [rspec](https://github.com/rspec/rspec)
- [capybara](https://github.com/jnicklas/capybara) and
  [database_cleaner](https://github.com/DatabaseCleaner/database_cleaner)
- [factory_girl](https://github.com/thoughtbot/factory_girl)
- [simplecov](https://github.com/colszowka/simplecov) for helping out with test
  coverage

Refer to the [rspec-rails](https://github.com/rspec/rspec-rails) to learn which
kinds of specs are available.

Note that you can require the following files to setup your tests:

- For light unit tests you can require `spec_helper.rb`. It won't boot up
  the Rails environment.
- For tests needing Rails you can require `rails_helper.rb`.
- For feature tests require `feature_helper.rb`. It will compile your assets
  and make feature tests run correctly.

Regarding feature tests, they are configured to run seamlessly with Webpack.

### Application server

We use [puma](https://github.com/puma/puma) as our application server, which
happens to be [Heroku's default
recommendation](https://devcenter.heroku.com/changelog-items/594).

### Background jobs

Our tool of choice is Sidekiq, which is configured as ActiveJob's backend. We
include a `sidekiq.yml` configuration file with default settings, but you are
encouraged to tune it to your application needs.

### Debugging

- [pry-rails](https://github.com/rweng/pry-rails)
- [pry-byebug](https://github.com/deivid-rodriguez/pry-byebug)
- [better-errors](https://github.com/charliesome/better_errors) instead of Web
  Console.

### Performance and Profiling

* [rack-mini-profiler](https://github.com/MiniProfiler/rack-mini-profiler) for
helping out with performance issues
* [spring](https://github.com/rails/spring) for fast Rails actions via
pre-loading
* [bullet](https://github.com/flyerhzm/bullet) yeah, it's very easy to miss out
N+1 queries, that's why we include this gem by default

Spring binstubs are automatically generated within the `bin` folder.

### Environment variables

* [Dotenv](https://github.com/bkeepers/dotenv) for loading environment variables

## Credits

The template uses the [Suspenders](https://github.com/thoughtbot/suspenders)
gem from Thoughtbot as starting point.

[![codeminer42](http://s3.amazonaws.com/cm42_prod/assets/logo-codeminer-horizontal-c0516b1b78a8713270da02c3a0645560.png)](http://codeminer42.com)


CHANGES.md
0.0.1 (January 28 2016)

* Initial setup
* Ruby 2.3.0
* Rails 4.2.5
* App layout and flashes
* Customize error pages
* Git init and .gitignore
* Postgresql as database
* pt-BR locale
* Readme
* Remove config and routes comment lines
* Setup script
* Setup secret token

1.0.0 (April 06 2016)

* Add `quiet_assets` as development dependency
* Generate rake, rails and rspec binstubs with Spring
* Make Shoulda Matchers work with Spring
* Add `binding_of_caller` gem
* Add `better_errors` gem
* Add `pry-byebug` and `pry-rails` gems
* Add `bullet` as development dependency
* Add Bundler Audit to scan Gemfile for insecure dependencies per CVEs.
* Add Dotenv to development and test environments to load environment variables from the `.env` file.
* Add `brakeman`
* Add `letter_opener`
* Add `rubocop`

1.0.1 (July 02 2016)

* Fix bug: remove missing jquery from generated application.js


asdf/config/application.rb
require_relative 'boot'

require "rails"
# Pick the frameworks you want:
require "active_model/railtie"
require "active_job/railtie"
require "active_record/railtie"
require "action_controller/railtie"
require "action_mailer/railtie"
require "action_view/railtie"
require "action_cable/engine"
require "sprockets/railtie"
# require "rails/test_unit/railtie"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module Asdf
  class Application < Rails::Application
    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

      # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
  # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
  config.time_zone = 'Madrid'


      # Locale
  # config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}')]
  # config.i18n.enforce_available_locales = false
  # config.i18n.default_locale = :es


      # Add to autoload_paths
  config.autoload_paths += %W{lib}

  end
end


asdf/config/environments/production.rb
Rails.application.configure do
  # Settings specified here will take precedence over those in config/application.rb.

  # Code is not reloaded between requests.
  config.cache_classes = true

  # Eager load code on boot. This eager loads most of Rails and
  # your application in memory, allowing both threaded web servers
  # and those relying on copy on write to perform better.
  # Rake tasks automatically ignore this option for performance.
  config.eager_load = true

  # Full error reports are disabled and caching is turned on.
  config.consider_all_requests_local       = false
  config.action_controller.perform_caching = true

  # Disable serving static files from the `/public` folder by default since
  # Apache or NGINX already handles this.
  config.public_file_server.enabled = ENV['RAILS_SERVE_STATIC_FILES'].present?

  # Compress JavaScripts and CSS.
  config.assets.js_compressor = :uglifier
  # config.assets.css_compressor = :sass

  # Do not fallback to assets pipeline if a precompiled asset is missed.
  config.assets.compile = false

  # `config.assets.precompile` and `config.assets.version` have moved to config/initializers/assets.rb

  # Enable serving of images, stylesheets, and JavaScripts from an asset server.
  # config.action_controller.asset_host = 'http://assets.example.com'

  # Specifies the header that your server uses for sending files.
  # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
  # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX

  # Mount Action Cable outside main process or domain
  # config.action_cable.mount_path = nil
  # config.action_cable.url = 'wss://example.com/cable'
  # config.action_cable.allowed_request_origins = [ 'http://example.com', /http:\/\/example.*/ ]

  # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
  # config.force_ssl = true

  # Use the lowest log level to ensure availability of diagnostic information
  # when problems arise.
  config.log_level = :debug

  # Prepend all log lines with the following tags.
  config.log_tags = [ :request_id ]

  # Use a different cache store in production.
  # config.cache_store = :mem_cache_store

  # Use a real queuing backend for Active Job (and separate queues per environment)
  # config.active_job.queue_adapter     = :resque
  # config.active_job.queue_name_prefix = "asdf_#{Rails.env}"
  config.action_mailer.perform_caching = false

  # Ignore bad email addresses and do not raise email delivery errors.
  # Set this to true and configure the email server for immediate delivery to raise delivery errors.
  # config.action_mailer.raise_delivery_errors = false

  # Enable locale fallbacks for I18n (makes lookups for any locale fall back to
  # the I18n.default_locale when a translation cannot be found).
  config.i18n.fallbacks = true

  # Send deprecation notices to registered listeners.
  config.active_support.deprecation = :notify

  # Use default logging formatter so that PID and timestamp are not suppressed.
  config.log_formatter = ::Logger::Formatter.new

  # Use a different logger for distributed setups.
  # require 'syslog/logger'
  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')

  if ENV["RAILS_LOG_TO_STDOUT"].present?
    logger           = ActiveSupport::Logger.new(STDOUT)
    logger.formatter = config.log_formatter
    config.logger = ActiveSupport::TaggedLogging.new(logger)
  end

  # Do not dump schema after migrations.
  config.active_record.dump_schema_after_migration = false
end


lib/cocodrilo/generators/app_generator.rb
require 'rails/generators'
require 'rails/generators/rails/app/app_generator'

module Cocodrilo
  class AppGenerator < Rails::Generators::AppGenerator
    class_option :database, type: :string, aliases: "-d", default: "postgresql",
      desc: "Configure for selected database (options: #{DATABASES.join("/")})"

    class_option :skip_test_unit, type: :boolean, aliases: "-T", default: true,
      desc: "Skip Test::Unit files"

    class_option :skip_turbolinks, type: :boolean, default: true,
      desc: "Skip turbolinks gem"

    class_option :skip_bundle, type: :boolean, aliases: "-B", default: true,
      desc: "Don't run bundle install"

      class_option :skip_test, type: :boolean, aliases: '-T', default: true,
        desc: 'Skip test files'

    def finish_template
      invoke :cocodrilo_customization
      super
    end

    def cocodrilo_customization
      invoke :customize_gemfile
      invoke :setup_application_rb
      invoke :setup_development_environment
      invoke :setup_test_environment
      invoke :setup_production_environment
      invoke :setup_secret_token
      invoke :configure_app
      invoke :setup_figaro
      invoke :setup_capistrano
      invoke :customize_error_pages
      invoke :setup_dotfiles
      invoke :setup_git
      invoke :setup_database
      invoke :setup_background_jobs
      invoke :setup_spring
      invoke :setup_health_task
    end

    def customize_gemfile
      build :set_ruby_to_version_being_used
      bundle_command 'install'
    end

    def setup_application_rb
      say 'Setting up application.rb'
      build :setup_time_zone
      build :setup_locale_configuration
      build :setup_autoload_paths
      # build :setup_lograge
      # build :setup_rack_cors
      # build :setup_generators
    end

    def setup_development_environment
      say 'Setting up the development environment'
      build :add_bullet_gem_configuration
      build :add_paperclip_gem_configuration_to_dev
      build :configure_quiet_assets
    end

    def setup_test_environment
      say 'Setting up the test environment'
      build :setup_test_env_action_dispatch_exceptions
    end

    def setup_production_environment
      say 'Setting up the production environment'
      build :add_paperclip_gem_configuration_to_prod
      build :add_rack_timeout
    end

    def setup_secret_token
      say 'Moving secret token out of version control'
    end

    def setup_figaro
      say 'Setting up figaro gem'
      build :application_yml
    end

    def setup_capistrano
      say 'Setting up capistrano'
      build :copy_capistrano_files
      build :copy_nginx_conf_files
    end

    def configure_app
      say 'Configuring app'
      build :add_puma_configuration
    end

    def customize_error_pages
      say 'Customizing the 500/404/422 pages'
      build :customize_error_pages
    end

    def setup_dotfiles
      build :copy_dotfiles
    end

    def setup_git
      if !options[:skip_git]
        say "Initializing git"
        invoke :init_git
      end
    end

    def init_git
      build :init_git
    end

    def setup_database
      return if options['skip_active_record']
      say 'Setting up database'
      build :use_postgres_config_template if 'postgresql' == options[:database]
      build :create_database
    end

    def setup_background_jobs
      say 'Setting up background jobs'
      build :setup_background_jogs
    end

    def setup_spring
      say "Springifying binstubs"
      build :setup_spring
    end

    def setup_health_task
      say "Setting up health task"
      build :setup_health_task
    end

    protected

    def get_builder_class
      Cocodrilo::AppBuilder
    end

    def using_active_record?
      !options[:skip_active_record]
    end
  end
end


lib/cocodrilo/app_builder.rb
require "forwardable"

module Cocodrilo
  class AppBuilder < Rails::AppBuilder
    include Cocodrilo::Actions
    extend Forwardable

    # Configuration files =================================

    def readme
      template 'README.md.erb', 'README.md'
    end

    def gitignore
      copy_file "cocodrilo_gitignore", ".gitignore"
    end

    def gemfile
      template "Gemfile.erb", "Gemfile"
    end

    def setup_secret_token
      template 'secrets.yml', 'config/secrets.yml', force: true
    end

    def application_yml
      copy_file "application.yml", "config/application.yml"
    end

    def copy_nginx_conf_files
      copy_file "nginx.conf.erb", "config/nginx.conf"
    end

    def copy_capistrano_files
      empty_directory 'config/deploy'
      copy_file "capistrano_production.rb", "config/deploy/production.rb"
      copy_file "capistrano_staging.rb", "config/deploy/staging.rb"
      copy_file "capistrano_deploy.rb.erb", "config/deploy.rb"
    end

    def add_puma_configuration
      copy_file "puma.rb", "config/puma.rb", force: true
    end

    def copy_dotfiles
      copy_file "dotfiles/rspec", ".rspec"
      copy_file "dotfiles/rubocop.yml", ".rubocop.yml"
    end

    def use_postgres_config_template
      template 'postgresql_database.yml.erb', 'config/database.yml',
        force: true
    end

    def setup_background_jobs
      copy_file 'active_job.rb', 'config/initializers/active_job.rb'
      copy_file 'sidekiq.yml', 'config/sidekiq.yml'
    end

    def set_ruby_to_version_being_used
      create_file '.ruby-version', "#{Cocodrilo::RUBY_VERSION}\n"
    end

    def configure_time_formats
      template "config_locales_en.yml.erb", "config/locales/en.yml"
    end

    def setup_health_task
      copy_file "health.rake", "lib/tasks/health.rake"
      append_file "Rakefile", %{\ntask default: "health"\n}
    end

    # Files injections ====================================

    def setup_time_zone
      config = <<RUBY
  # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
  # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
  config.time_zone = 'Madrid'
RUBY
      configure_application_file(config)
    end

    def setup_locale_configuration
      config = <<RUBY
  # Locale
  # config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}')]
  # config.i18n.enforce_available_locales = false
  # config.i18n.default_locale = :es
RUBY
      configure_application_file(config)
    end

    def setup_autoload_paths
      config = <<RUBY
  # Add to autoload_paths
  config.autoload_paths += %W{lib}
RUBY
      configure_application_file(config)
    end

    def setup_lograge
      config = <<RUBY
  # Logs
  unless Rails.env.test?
    log_level = String(ENV['LOG_LEVEL']||"info").upcase
    config.log_level = log_level
    config.lograge.enabled = true

    INTERNAL_PARAMS = %w(controller action format _method only_path)

    config.lograge.custom_options = lambda do |event|
      payload = event.payload
      params  = payload[:params].except(*INTERNAL_PARAMS)
      host    = payload[:host]
      ip      = payload[:ip]
      { params: params, host: host, ip: ip, time: event.time }
    end
  end
RUBY
      configure_application_file(config)
    end

    def setup_rack_cors
      config = <<RUBY
  # Rack::Cors configuration
  config.middleware.insert_before 0, Rack::Cors do
    allow do
      origins '*'
      resource '*', :headers => :any, :methods => [:get, :post, :options]
    end
  end
RUBY

      configure_application_file(config)
    end

    def setup_generators
      config = <<RUBY
  # Generators
  config.generators do |g|
    g.helper false
    g.javascript_engine false
    g.stylesheets false
    g.request_specs false
    g.routing_specs false
    g.view_specs false
    g.test_framework nil
    # g.test_framework :rspec
    #                  :fixtures => false,
    #                  :model_specs => true,
    #                  :view_specs => false,
    #                  :helper_specs => false,
    #                  :routing_specs => false,
    #                  :controller_specs => true,
    #                  :request_specs => false
  end
RUBY

      configure_application_file(config)
    end

    def add_bullet_gem_configuration
      config = <<RUBY
  config.after_initialize do
    Bullet.enable = true
    Bullet.bullet_logger = true
    Bullet.rails_logger = true
  end
RUBY

      inject_into_file(
        "config/environments/development.rb",
        config,
        after: "config.file_watcher = ActiveSupport::EventedFileUpdateChecker\n",
      )
    end

    def add_paperclip_gem_configuration_to_dev
      config = <<RUBY
  config.paperclip_defaults = {
    :storage => :s3,
    :s3_region => ENV['AWS_REGION'],
    :s3_credentials => {
      :bucket => ENV['S3_BUCKET_NAME'],
      :access_key_id => ENV['AWS_ACCESS_KEY_ID'],
      :secret_access_key => ENV['AWS_SECRET_ACCESS_KEY']
    }
  }
  end
RUBY

      inject_into_file(
        "config/environments/development.rb",
        config,
        after: "config.file_watcher = ActiveSupport::EventedFileUpdateChecker\n",
      )
    end

    def add_paperclip_gem_configuration_to_prod
      config = <<RUBY
  config.paperclip_defaults = {
    :storage => :s3,
    :s3_region => ENV['AWS_REGION'],
    :s3_credentials => {
      :bucket => ENV['S3_BUCKET_NAME'],
      :access_key_id => ENV['AWS_ACCESS_KEY_ID'],
      :secret_access_key => ENV['AWS_SECRET_ACCESS_KEY']
    }
  }
  end
RUBY

      inject_into_file(
        "config/environments/production.rb",
        config,
        after: "config.active_record.dump_schema_after_migration = false\n",
      )
    end

    def add_rack_timeout
      config = "Rack::Timeout.timeout = (ENV['RACK_TIMEOUT'] || 10).to_i"
      inject_into_file(
        "config/environments/production.rb",
        config,
        after: "config.active_record.dump_schema_after_migration = false\n",
      )
    end

    def customize_error_pages
      meta_tags =<<-EOS
  <meta charset="utf-8" />
  <meta name="ROBOTS" content="NOODP" />
  <meta name="viewport" content="initial-scale=1" />
      EOS

      %w(500 404 422).each do |page|
        inject_into_file "public/#{page}.html", meta_tags, after: "<head>\n"
        replace_in_file "public/#{page}.html", /<!--.+-->\n/, ''
      end
    end

    def setup_test_env_action_dispatch_exceptions
      gsub_file(
        'config/environments/test.rb',
        'config.action_dispatch.show_exceptions = false',
        'config.action_dispatch.show_exceptions = true'
      )
    end

    def configure_quiet_assets
      config = <<-RUBY
    config.assets.quiet = true
      RUBY

      inject_into_class "config/application.rb", "Application", config
    end

    # Execute commands ====================================

    def create_database
      bundle_command 'exec rake db:create db:migrate'
    end

    def setup_spring
      bundle_command "exec spring binstub --all"
    end

    # Git =================================================

    def init_git
      run 'git init'
    end

  end
end
